app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\LeadBundle\EventListener;
+
+use Mautic\CampaignBundle\CampaignEvents;
+use Mautic\CampaignBundle\Event\CampaignBuilderEvent;
+use Mautic\CampaignBundle\Event\PendingEvent;
+use Mautic\CoreBundle\Model\AuditLogModel;
+use Mautic\LeadBundle\Entity\Company;
+use Mautic\LeadBundle\Entity\Lead;
+use Mautic\LeadBundle\Entity\LeadField;
+use Mautic\LeadBundle\Form\Type\CampaignActionAnonymizeUserDataType;
+use Mautic\LeadBundle\LeadEvents;
+use Mautic\LeadBundle\Model\CompanyModel;
+use Mautic\LeadBundle\Model\FieldModel;
+use Mautic\LeadBundle\Model\LeadModel;
+use Mautic\LeadBundle\Services\AnonymizeContactCompanyData;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class CampaignActionAnonymizeUserDataSubscriber implements EventSubscriberInterface
+{
+    public const KEY_EVENT_NAME = 'lead.action_anonymizeuserdata';
+
+    public function __construct(
+        private LeadModel $leadModel,
+        private FieldModel $fieldModel,
+        private CompanyModel $companyModel,
+        private AnonymizeContactCompanyData $anonymizeContactCompanyData,
+        private AuditLogModel $auditLogModel,
+    ) {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CampaignEvents::CAMPAIGN_ON_BUILD                  => ['configureAction', 0],
+            LeadEvents::ON_CAMPAIGN_ACTION_ANONYMIZE_USER_DATA => ['anonymizeUserData', 10],
+        ];
+    }
+
+    public function configureAction(CampaignBuilderEvent $event): void
+    {
+        $event->addAction(
+            self::KEY_EVENT_NAME,
+            [
+                'label'                  => 'mautic.lead.lead.events.anonymize',
+                'description'            => 'mautic.lead.lead.events.anonymize_descr',
+                // Kept for BC in case plugins are listening to the shared trigger
+                'eventName'              => LeadEvents::ON_CAMPAIGN_TRIGGER_ACTION, // @phpstan-ignore-line
+                'formType'               => CampaignActionAnonymizeUserDataType::class,
+                'batchEventName'         => LeadEvents::ON_CAMPAIGN_ACTION_ANONYMIZE_USER_DATA,
+            ]
+        );
+    }
+
+    public function anonymizeUserData(PendingEvent $event): void
+    {
+        if (!$event->checkContext(self::KEY_EVENT_NAME)) {
+            return;
+        }
+
+        $properties           = $event->getEvent()->getProperties();
+        $isToPseudonymize     = false;
+        if (array_key_exists('pseudonymize', $properties)) {
+            $isToPseudonymize = (bool) $properties['pseudonymize'];
+        }
+        $leadFilter           = [
+            'force' => [
+                ['column' => 'l.id', 'expr' => 'in', 'value' => $event->getContactIds()],
+            ],
+        ];
+        $leads            = $this->leadModel->getEntities($leadFilter);
```

Comment:
We do not need get entities, we already have in PendingEvent logs and of each log has lead entity.